compiler/rustc_target/src/callconv/wasm.rs

```diff
@@ -27,6 +29,16 @@ where
     if ret.layout.is_aggregate() && !unwrap_trivial_aggregate(cx, ret) {
         ret.make_indirect();
     }
+
+    // `long double`, `__int128_t` and `__uint128_t` use an indirect return
+    if let BackendRepr::Scalar(scalar) = ret.layout.backend_repr {
+        match scalar.primitive() {
+            Primitive::Int(Integer::I128, _) | Primitive::Float(Float::F128) => {
+                ret.make_indirect();
+            }
+            _ => {}
+        }
+    }
```

Comment:
Could remove some nesting

```suggestion
    if matches!(
        ret.layout.backend_repr,
        BackendRepr::Scalar(Primitive::Int(Integer::I128, _) | Primitive::Float(Float::F128))
    ) {
                ret.make_indirect();
    }
```

Reply:
unless there is some magic going on here, the `BackendRepr::Scalar` contains a `Scalar`, not a `Primitive`. `Scalar` does not seem fun to pattern match on.

Reply:
Nope you are right, missed that extra level.